### PR TITLE
Stop logging retry messages when retries won't actually be performed

### PR DESCRIPTION
**Pull request by nathanwoctopusdeploy**
# Background

Don't log that we will retry when there is no time left and reties won't be performed.
Don't log that retries have timed out if we never performed them as there was no time e.g. the initial request took longer than the retry duration.

## Before

Would log that a retry was going to be performed when there was no time left.

image

Would log that retries were attempted even if none were performed. 

## After

Will not log we are about to retry when there is no time left.
Will not log that retries timed out if no retries were attempted.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs [hunk 1]
```diff
@@ -41,21 +41,31 @@ public RpcCallRetryHandler(TimeSpan retryTimeout, TimeoutStrategy timeoutStrateg
...
+                    {
+                        await onRetryAction.Invoke(exception, sleepDuration, retryCount, RetryTimeout, elapsedDuration, cancellationToken).ConfigureAwait(false);
+                    }
                 }
             }
 
             async Task OnTimeoutAction(Context? context, TimeSpan timeout, Task? task, Exception? exception)
             {
-                if (onTimeoutAction != null)
+                if (hasRetried && onTimeoutAction != null)
```
Comment (acodrington) [inline source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs:68]: Why do we need to check `hasRetried` here? Wouldn't we still want to call the timeout action if the original request (before retries) times out?

Comment (nathanwoctopusdeploy) [reply source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs:68]: Yerp, I've done this to stop logging a misleading message to the user "Could not communicate with Teantcale after 150 seconds. No more retries will be attempted."

As we probably took more than 150 seconds in this scenario as we are still on the initial request and never retried, and saying no more retries will be attempted sounds like we attempted retries that we didn't!!

But as you point out it seems sus to not call on timeout!! I'll look at a better way of logging what really happened to the end user!

File: source/Octopus.Tentacle.Client/Execution/RpcCallExecutor.cs [hunk 2]
```diff
@@ -63,14 +63,24 @@ internal class RpcCallExecutor
...
+                            logger.Info($"An error occurred communicating with Tentacle. This action will be retried after {(int)sleepDuration.TotalSeconds} seconds. Retry attempt {retryCount}. Retries will be performed for up to {remainingDurationInSeconds} seconds.");
                             logger.Verbose(lastException);
                         },
-                        onTimeoutAction: async (timeoutDuration, _) =>
+                        onTimeoutAction: async (timeoutDuration, elapsedDuration, retryCount, _) =>
                         {
                             await Task.CompletedTask;
-                            logger.Info($"Could not communicate with Tentacle after {timeoutDuration.TotalSeconds} seconds. No more retries will be attempted.");
+
+                            if (retryCount > 0)
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Client/Execution/RpcCallExecutor.cs:76]: This is the change!!

File: source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs [hunk 3]
```diff
@@ -41,23 +43,34 @@ public RpcCallRetryHandler(TimeSpan retryTimeout, TimeoutStrategy timeoutStrateg
...
             var started = new Stopwatch();
+            var nextSleepDuration = TimeSpan.Zero;
+            var totalRetryCount = 0;
 
             async Task OnRetryAction(Exception exception, TimeSpan sleepDuration, int retryCount, Context context)
             {
                 lastException = exception;
+                nextSleepDuration = sleepDuration;
                 var elapsedDuration = started.Elapsed;
+                var remainingRetryDuration = RetryTimeout - elapsedDuration - sleepDuration;
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs:54]: We didn't take into account the sleep duration so often thought we would retry when we had actually run out of time

File: source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs [hunk 4]
```diff
@@ -41,23 +43,34 @@ public RpcCallRetryHandler(TimeSpan retryTimeout, TimeoutStrategy timeoutStrateg
...
 
-                if (onRetryAction != null)
+                if (ShouldRetryWithRemainingDuration(remainingRetryDuration))
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Client/Retries/RpcCallRetryHandler.cs:56]: We would log the retry messages even when there was no time to do retries

File: source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs [hunk 5]
```diff
@@ -83,7 +88,51 @@ public async Task WhenRpcRetriesTimeOut_DuringUploadFile_TheRpcCallIsCancelled(T
             fileTransferServiceCallCounts.DownloadFileCallCountStarted.Should().Be(0);
 
             // Ensure we actually waited and retried until the timeout policy kicked in
-            duration.Elapsed.Should().BeGreaterOrEqualTo(TimeSpan.FromSeconds(14));
+            duration.Elapsed.Should().BeGreaterOrEqualTo(clientAndTentacle.RpcRetrySettings.RetryDuration - retryIfRemainingDurationAtLeastBuffer - retryBackoffBuffer);
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs:91]: Make tests more reliable

File: source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs [hunk 6]
```diff
@@ -83,7 +88,51 @@ public async Task WhenRpcRetriesTimeOut_DuringUploadFile_TheRpcCallIsCancelled(T
...
+
+            RetryLogMessageAssertions.AssertRetryAttemptsLoggedAndRetryFailureLogged(inMemoryLog);
+        }
+
+        [Test]
+        [TentacleConfigurations]
+        public async Task WhenUploadFileFails_AndTakesLongerThanTheRetryDuration_TheCallIsNotRetried_AndTimesOut(TentacleConfigurationTestCase tentacleConfigurationTestCase)
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs:98]: We had a gap in test coverage when the initial request took longer than the allowed retry duration

File: source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs [hunk 7]
```diff
@@ -83,7 +88,51 @@ public async Task WhenRpcRetriesTimeOut_DuringUploadFile_TheRpcCallIsCancelled(T
             fileTransferServiceCallCounts.DownloadFileCallCountStarted.Should().Be(0);
 
             // Ensure we actually waited and retried until the timeout policy kicked in
-            duration.Elapsed.Should().BeGreaterOrEqualTo(TimeSpan.FromSeconds(14));
+            duration.Elapsed.Should().BeGreaterOrEqualTo(clientAndTentacle.RpcRetrySettings.RetryDuration - retryIfRemainingDurationAtLeastBuffer - retryBackoffBuffer);
+
+            RetryLogMessageAssertions.AssertRetryAttemptsLoggedAndRetryFailureLogged(inMemoryLog);
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRetriesTimeout.cs:93]: Assertions have been added to the existing tests to cover the task log output

File: source/Octopus.Tentacle.Tests.Integration/ClientGathersRpcCallMetrics.cs [hunk 8]
```diff
@@ -204,7 +204,7 @@ private static void ThenClientOperationMetricsShouldBeSuccessful(ClientOperation
             metric.Exception.Should().BeNull();
             metric.WasCancelled.Should().BeFalse();
 
-            metric.End.Should().BeAfter(metric.Start);
+            metric.End.Should().BeOnOrAfter(metric.Start);
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/ClientGathersRpcCallMetrics.cs:207]: This was showing up as an unrelated flakey test

File: source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionRetriesTimeout.cs [hunk 9]
```diff
@@ -98,7 +103,66 @@ await syncOrAsyncHalibut.WhenSync(() => scriptServiceV2.EnsureTentacleIsConnecte
...
             // Ensure we actually waited and retried until the timeout policy kicked in
-            duration.Elapsed.Should().BeGreaterOrEqualTo(TimeSpan.FromSeconds(14));
+            duration.Elapsed.Should().BeGreaterOrEqualTo(clientAndTentacle.RpcRetrySettings.RetryDuration - retryIfRemainingDurationAtLeastBuffer - retryBackoffBuffer);
+
+            RetryLogMessageAssertions.AssertRetryAttemptsLoggedAndRetryFailureLogged(inMemoryLog);
+        }
+
+        [Test]
+        [TentacleConfigurations]
+        public async Task WhenGetCapabilitiesFails_AndTakesLongerThanTheRetryDuration_TheCallIsNotRetried_AndTimesOut(TentacleConfigurationTestCase tentacleConfigurationTestCase)
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/ClientScriptExecutionRetriesTimeout.cs:113]: Additional testing to cover the scenario where the initial request takes longer than the retry duration

File: source/Octopus.Tentacle.Tests.Integration/Octopus.Tentacle.Tests.Integration.csproj [hunk 10]
```diff
@@ -32,7 +32,7 @@
         <PackageReference Include="FluentAssertions" Version="6.7.0" />
         <PackageReference Include="Assent" Version="1.8.2" />
         <PackageReference Include="NUnit" Version="3.13.3" />
-        <PackageReference Include="Octopus.TestPortForwarder" Version="6.0.1069" />
+        <PackageReference Include="Octopus.TestPortForwarder" Version="6.0.1415" />
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/Octopus.Tentacle.Tests.Integration.csproj:35]: Bump as some improvements have been made to disposal of the port forwarder

File: source/Octopus.Tentacle.Tests.Integration/Support/IntegrationTest.cs [hunk 11]
```diff
@@ -1,5 +1,4 @@
 ﻿using System;
-using System.Diagnostics;
 using System.Threading;
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/Support/IntegrationTest.cs:2]: Simplification of this class to align with the implementation in Halibut and remove some possible issues our test infra creates

File: source/Octopus.Tentacle.Tests.Integration/Support/SetupFixtures/BumpThreadPoolForAllTests.cs [hunk 12]
```diff
@@ -9,8 +9,8 @@ public class BumpThreadPoolForAllTests : ISetupFixture
         public void OneTimeSetUp(ILogger logger)
         {
             logger.Information("Bumping thread pool");
-            var minWorkerPoolThreads = 5000;
-            var minCompletionPortThreads = 5000;
+            var minWorkerPoolThreads = 1000;
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/Support/SetupFixtures/BumpThreadPoolForAllTests.cs:12]: We have observed in testing that having a min thread pool that is too large can hurt performance as .net struggles to manage the scheduling of the threads.

I observed similar issues locally where a test would just freeze for a long duration and then fail. Reducing this as a possible contributor

File: source/Octopus.Tentacle.Tests/Client/DefaultScriptObserverBackoffStrategyTests.cs [hunk 13]
```diff
@@ -7,15 +7,9 @@
 namespace Octopus.Tentacle.Tests.Client
 {
     [TestFixture]
+    [Parallelizable(ParallelScope.All)]
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests/Client/DefaultScriptObserverBackoffStrategyTests.cs:10]: By default unit tests run sequentially but when safe to do so we should allow them to run in parallel. Some tests are not safe to do this for so we cannot do a blanket opt-in. Instead we can opt-in when we know it's safe to do.

**Comment by nathanwoctopusdeploy** [comment]
[sc-58644]

File: source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        public static void AssertRetryAttemptsLoggedButNoRetryFailureLogged(InMemoryLog logs)
+        {
+            AssertRetriedLogMessages(logs);
+            
+            // Negative assertions
+            AssertNoTimeoutAfterRetriedLogMessages(logs);
+            AssertNoTimeoutAfterNoRetriedLogMessages(logs);
+        }
+
+        public static  void AssertRetryAttemptsLoggedAndRetryFailureLogged(InMemoryLog logs)
```
Comment (sburmanoctopus) [inline source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs:21]: Nit: Double spaces? (further down too)

File: source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            AssertRetriedLogMessages(logs);
+            AssertTimeoutAfterRetriedLogMessages(logs);
+
+            // Negative assertions
+            AssertNoTimeoutAfterNoRetriedLogMessages(logs);
+        }
+
+        public static  void AssertNoRetryAttemptsLoggedAndRetryFailureLogged(InMemoryLog logs)
```
Comment (sburmanoctopus) [inline source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs:30]: Just a thought, but we could make this an extension method, and even make the method names match the fluent "Should" style.

Then this:
```
RetryLogMessageAssertions.AssertRetryAttemptsLoggedAndRetryFailureLogged(inMemoryLog);
```

Could be something like this:
```
inMemoryLog.ShouldHaveLoggedRetryAttemptsAndRetryFailure();
```

**Review by sburmanoctopus — APPROVED** [review]
LGTM

File: source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs (hunk 15 above)
Comment (nathanwoctopusdeploy) [reply source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs:30]: Yeh nice, I'll rework this

File: source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs (hunk 14 above)
Comment (nathanwoctopusdeploy) [reply source/Octopus.Tentacle.Tests.Integration/Support/RetryLogMessageAssertions.cs:21]: Impressive spot 😆

**Pull request merged**